### ISSUE TITLE
HDK Wish list

### ISSUE THREAD

**Issue by guillemcordoba** [holochain/holochain#889]
## Pretty much needed

- [x] Remove unnecessary verbose types:
  - https://docs.rs/hdk/0.0.101/hdk/prelude/struct.Links.html
- [x] Ability to get the provenance with maybe agent_info? (agent pub key that called the executing function).
- [x] Ability to easily `call` another cell in the same happ referring to it by its nick.
  - Right now I have to [hardcode the hash of the DNA in the other DNA's properties](https://github.com/holochain-open-dev/file-storage/blob/7f77a5033f4d05f7bb4c45711dfd595091e8c293/sample-happ/dnas/consumer/workdir/dna.yaml#L6), not ideal and will break with change of UID or properties.
- [x] [Sleep](https://docs.rs/hdk/0.0.101/hdk/time/fn.sleep.html). // For now, this may be covered by `schedule()`

## Pretty please? :)
- [ ] Improve developer experience of converting an arbitrary type from and to `LinkTag` (with `SerializedBytes`?). #2399 
  -  [This is how it is right now](https://github.com/holochain-open-dev/geolocation/blob/ #ee8d240d01303027ba80b74fd560d52406c5be7d/zome/zomes/geolocation/src/lib.rs#L44), a bit confusing.
- [ ] #4422
- [x] Ability to get the names of different zomes in the DNA.
- [x] Right now it's [pretty verbose](https://github.com/holochain/syn/blob/99ed64adedee4b8a0e35f20e89e83f9ecce0af8c/zomes/syn/src/session.rs#L58) to have a workable timestamp from `sys_time`'s `Duration`.
- [x] Remove [Links](https://docs.rs/hdk/0.0.103/hdk/prelude/struct.Links.html) type, it seems a leftover from the days in which only types with `SerializedBytes` could pass the wasm boundary, right now I think it only adds complexity, it's a bit confusing for new devs until you find `.into_inner()`.
- [x] Remove [HeaderHashes](https://docs.rs/hdk/0.0.110/hdk/prelude/struct.HeaderHashes.html) and [HeaderHashedVec](https://docs.rs/hdk/0.0.110/hdk/prelude/struct.HeaderHashedVec.html), same reason as with `Links`.
- [x] Enable Get Agent Activity in validation rules. This has been discussed with @thedavidmeister as changing the input of that function to a range of header hashes to make it deterministic.
- [x] Add the author of the link to the link struct.

## Other ideas

- [ ] host fn builder to batch heterogeneous fn calls

**Comment by tatssato** [comment]
## Pretty much needed
- [ ] post_commit (interested in helping)

## Pretty please? :)
- [ ] ability to query source chain for a specific Entry in constant time? (e.g. query by EntryHash)... might be shooting for a moon here

**Comment by artbrock** [comment]
Just some quick responses to some of these items from @guillemcordoba & @tatssato 

## Pretty Much Needed Items
Yes... to all of these with one caveat,  I think it will be confusing to have caller info in the agent_info (which is about who YOU are). However, I've documented a `call_info()` which let's you inspect the context the current running zome call was made including:
 - source -- network, client API, local cell, cross-zome (within current DNA), etc.
 - provenance -- calling agent info
 - permitted functions -- the functions unlocked by the cap token
 - captoken hash -- which captoken is currently in play (maybe?? -- I'm not sure wasm should be able to access this because it could leak secrets in the captokens)

**Comment by timotree3** [comment]
# Pretty please
- Find out own source chain top hash in constant time (Current only way is to call `hdk::chain::query()` with no filter)
- Create links with the base or target or both on a header hash instead of entry hash

**Comment by zippy** [comment]
## Request

Be able to get the UID back from AppInfo call.  Maybe this is what distinguishes AppInfo from ZomeInfo?

**Comment by thedavidmeister** [comment]
maybe this https://github.com/holochain/holochain/issues/743#issuecomment-896446332

**Comment by thedavidmeister** [comment]
and this https://github.com/holochain/holochain/issues/976

**Comment by thedavidmeister** [comment]
and this https://github.com/holochain/holochain/issues/339

**Comment by thedavidmeister** [comment]
`Timestamp` has been heavily refactored and standardised by @maackle and also `sys_time` returns a `Timestamp` instead of a `Duration` so we don't need to move between the two so much, which is awkward anyway as the former is an absolute time and the latter is an amount of time - there's no way to automatically move between the two without making some big assumptions about what the `Duration` is relative to

**Comment by thedavidmeister** [comment]
at least a test for https://github.com/holochain/holochain/issues/693

**Comment by thedavidmeister** [comment]
- [ ] another common request is that permissions issues should not panic either

**Comment by nphias** [comment]
- [ ] new structure for AppSignal to indicate the zome that the signal came from . 

Instead of this:
```
 AppSignal = {
    type: string;
    data: {
        cellId: CellId;
        payload: any;
    };
}; 
```
this:
```
AppSignal = {
    type: string;
    data: {
        cellId: CellId;
        zome_name: string
        payload: any;
    };
};
```

- [ ] Improved documentation on what the `type` field represents..
- [ ] An expected `Payload` struct type in rust for zomes `{ name:string, data:any} `where we can get the signal name and than arbitrary data with any structure

**Comment by zippy** [comment]
[ ] an `as_ref_mut` accessor to the `Component` list in a `Path`?

**Comment by guillemcordoba** [comment]
[ ] re-export `holo_hash` with the "encoding" feature so we don't have to keep enabling it in our Cargo.toml if we want to use B64 hashes: 
```toml
[dependencies]
derive_more = "0"
serde = "1"

hdk = "0.0.115"
holo_hash = {version = "0.0.12", features = ["encoding"]}
```

**Comment by pdaoust** [comment]
[ ] allow local signals to be emitted from `post_commit` callback

**Comment by matthme** [comment]
- [ ] have `dna_info()` return also a "base dna hash", i.e. a dna hash generated from the dna without or with default properties.

I think this would greatly improve implementation of cross-cell UI's in that it would be possible for UI's to detect compatible cells within the conductor.
Concrete example: A calendar UI can easily ask the conductor for all dna's of one or multiple compatible base dna hash(es), although they each have a different network seed, and render all calendar events from all compatible cells.

**Comment by ThetaSinner** [comment]
@pdaoust @mattyg For the ergonomics and what might already done, could you take a look at the remaining items on here?

**Comment by mattyg** [comment]
> [ ] an `as_ref_mut` accessor to the `Vec<Component>` list in a `Path` ( this allows it to be used with vec `append`)

@zippy Do you think this is addressed by the the `Path` function `.append_component()`? https://docs.rs/hdi/latest/hdi/hash_path/path/struct.Path.html#method.append_component

**Comment by mattyg** [comment]
>  Improved documentation on what the type field represents..
@nphias Do you know if this is still relevant? I can't find the type field in AppSignal, so maybe it has been removed?

**Comment by mattyg** [comment]
I reviewed all the requests here and confirmed that they have either been completed, or have their own standalone issues.

I created standalone issues for those missing: #4429 #4426 #4428 #4427 #4425.

A few of the requests I believe are no longer relevant but I'm not certain, so I tagged the original requester in comments here asking for clarification.

Closing this ticket in favor of standalone issues.
